### PR TITLE
sap_general_preconfigure: Use tags for limiting the role scope

### PR DESCRIPTION
**Pull request by berndfinger**
Please see the updates of the `README.md` file for examples for using tags with this role (currently only for RHEL systems).
Solves #342 but leaves the related role parameters (e.g. `sap_general_preconfigure_configuration`, `sap_general_preconfigure_2772999_02`) in place and functional because no deprecation note has been published yet.

### CODE REVIEW

File: roles/sap_general_preconfigure/tasks/RedHat/configuration.yml [hunk 1]
```diff
@@ -35,7 +35,11 @@
...
 - name: Configure - Include configuration actions for required sapnotes
-  ansible.builtin.include_tasks: "sapnote/{{ sap_note_line_item.number }}.yml"
+  ansible.builtin.include_tasks:
+    file: "sapnote/{{ sap_note_line_item.number }}.yml"
   with_items: "{{ __sap_general_preconfigure_sapnotes_versions | difference(['']) }}"
   loop_control:
     loop_var: sap_note_line_item
+  tags:
+    - sap_general_preconfigure_configuration
+    - sap_general_preconfigure_configuration_all_steps
```
Comment (Wabri) [inline roles/sap_general_preconfigure/tasks/RedHat/configuration.yml:45]: Is the tag `sap_general_preconfigure_configuration_all_steps` really necessary?

File: roles/sap_general_preconfigure/tasks/main.yml [hunk 2]
```diff
@@ -52,45 +61,75 @@
...
   ansible.builtin.package_facts:
+  tags:
+    - sap_general_preconfigure_installation
 
 - name: Include tasks from 'installation.yml'
-  ansible.builtin.include_tasks: '{{ item }}/{{ __sap_general_preconfigure_fact_assert_filename_prefix }}installation.yml'
+  ansible.builtin.include_tasks:
+    file: '{{ item }}/{{ __sap_general_preconfigure_fact_assert_filename_prefix }}installation.yml'
+    apply:
+      tags: sap_general_preconfigure_installation
```
Comment (Wabri) [inline roles/sap_general_preconfigure/tasks/main.yml:102]: There is already a tag in this task, why adding the same one?

File: roles/sap_general_preconfigure/tasks/main.yml [hunk 3]
```diff
@@ -52,45 +61,75 @@
...
+  tags:
+    - sap_general_preconfigure_installation
+    - sap_general_preconfigure_configuration
 
 - name: Include tasks from 'configuration.yml'
-  ansible.builtin.include_tasks: '{{ item }}/{{ __sap_general_preconfigure_fact_assert_filename_prefix }}configuration.yml'
+  ansible.builtin.include_tasks:
+    file: '{{ item }}/{{ __sap_general_preconfigure_fact_assert_filename_prefix }}configuration.yml'
+    apply:
+      tags: sap_general_preconfigure_configuration_all_steps
```
Comment (Wabri) [inline roles/sap_general_preconfigure/tasks/main.yml:121]: Same question done in the line 102

File: roles/sap_general_preconfigure/tasks/RedHat/configuration.yml (hunk 1 above)
Comment (berndfinger) [reply roles/sap_general_preconfigure/tasks/RedHat/configuration.yml:45]: > Is the tag `sap_general_preconfigure_configuration_all_steps` really necessary?

Thanks! I tested again and it found no longer a combination of tags which would require both tags. So it seems it's really not required. I'll update the code accordingly.

File: roles/sap_general_preconfigure/tasks/main.yml (hunk 2 above)
Comment (berndfinger) [reply roles/sap_general_preconfigure/tasks/main.yml:102]: This code will apply a tag to all tasks of the included file(s). And without the tags: line below, in the same task, this current task will not be run when using the tag. See also: https://docs.ansible.com/ansible/latest/collections/ansible/builtin/include_tasks_module.html .

File: roles/sap_general_preconfigure/tasks/main.yml (hunk 3 above)
Comment (berndfinger) [reply roles/sap_general_preconfigure/tasks/main.yml:121]: See above.

File: roles/sap_general_preconfigure/tasks/RedHat/configuration.yml (hunk 1 above)
Comment (berndfinger) [reply roles/sap_general_preconfigure/tasks/RedHat/configuration.yml:45]: I was a bit too fast. Further testing revealed that only with both tags, `sap_general_preconfigure_configuration` and  `sap_general_preconfigure_configuration_all_steps`, it is possible to execute just one or a few of the configuration steps.
With just the tag `sap_general_preconfigure_configuration`, it is possible to run all configuration steps and skip some of them. But I want more. ;-)
See also the examples in README.md.

File: roles/sap_general_preconfigure/tasks/main.yml (hunk 2 above)
Comment (rhmk) [reply roles/sap_general_preconfigure/tasks/main.yml:102]: LGTM

**Review by rhmk — APPROVED** [review]
The changes look good to me and really improve usability

**Review by Wabri — APPROVED** [review]
LGTM

**Comment by berndfinger** [comment]
After discussing an alternative solution with a colleague (for avoiding two parameters when running all or most configuration steps), I will test another approach. Moving to draft now.

**Comment by berndfinger** [comment]
This version is now easier to use and eliminates the need for specifying two tags when running all configuration steps. The only little disadvantage is that now, when using the tag `sap_general_preconfigure_configuration` only, also some header lines of configuration tasks (displaying the SAP note numbers and versions) are displayed. But as using the tags should not be a default, I think this can be accepted.
See the additional lines in `README.md` for examples. Thanks to @Wabri and @ja9fuchs for triggering this change!

**Comment by berndfinger** [comment]
BTW We now know that https://github.com/sap-linuxlab/community.sap_install/pull/652 (ansible-test-sanity for every PR) works as expected. Run time was 1m 32s. Thanks @Wabri !

**Comment by rainerleber** [comment]
LGTM. This PR is a really cool improvement.

**Review by rainerleber — APPROVED** [review]

File: roles/sap_general_preconfigure/tasks/sapnote/2002167.yml
Comment (ja9fuchs) [inline roles/sap_general_preconfigure/tasks/sapnote/2002167.yml:1]: Would it make sense to apply a tag of `sap_general_preconfigure_2002167` to all the imported task files of this specific note? 
As I understand it right now, applying only all tasks for 2002167 requires to list 5 tags. Is that intended?

File: roles/sap_general_preconfigure/tasks/sapnote/2772999.yml
Comment (ja9fuchs) [inline roles/sap_general_preconfigure/tasks/sapnote/2772999.yml:1]: Same as above, what is the chance of a use-case to only apply all steps related to note 277999? If this is a realistic scenario I suggest to add a tag for the entire note to all the related imported tasks.

File: roles/sap_general_preconfigure/tasks/sapnote/assert-2772999.yml [hunk 4]
```diff
@@ -4,39 +4,59 @@
...
   ansible.builtin.import_tasks: 2772999/08-assert-linux-kernel-parameters.yml
   when: sap_general_preconfigure_config_all | d(true) or sap_general_preconfigure_2772999_08 | d(false)
+  tags:
+    - sap_general_preconfigure_2772999_08
 
 - name: Import tasks from '2772999/09-assert-process-resource-limits.yml'
   ansible.builtin.import_tasks: 2772999/09-assert-process-resource-limits.yml
   when: sap_general_preconfigure_config_all | d(true) or sap_general_preconfigure_2772999_09 | d(false)
+  tags:
+    - sap_general_preconfigure_nproc_limits
```
Comment (ja9fuchs) [inline roles/sap_general_preconfigure/tasks/sapnote/assert-2772999.yml:56]: I'm a little bit confused here by the inconsistency. Why not (also) add a tag `sap_general_preconfigure_2772999_09` to this task, similar to the others? And why are there no "theme" tags like `..._kernel_parameters`, etc., here, similar to the other note tasks?

**Review by ja9fuchs — CHANGES_REQUESTED** [review]

File: roles/sap_general_preconfigure/tasks/sapnote/2002167.yml
Comment (berndfinger) [reply roles/sap_general_preconfigure/tasks/sapnote/2002167.yml:1]: Thanks for the idea! With this mechanism we can simplify implementing just one of the notes! I will implement and test and then update this PR.

File: roles/sap_general_preconfigure/tasks/sapnote/2772999.yml
Comment (berndfinger) [reply roles/sap_general_preconfigure/tasks/sapnote/2772999.yml:1]: (See above)

File: roles/sap_general_preconfigure/tasks/sapnote/assert-2772999.yml (hunk 4 above)
Comment (berndfinger) [reply roles/sap_general_preconfigure/tasks/sapnote/assert-2772999.yml:56]: This is a coding error which I need to fix.

Comment (berndfinger) [reply roles/sap_general_preconfigure/tasks/sapnote/assert-2772999.yml]: @ja9fuchs BTW The coding error only affects line 56. Regarding the "theme" tags, I am defining those in the files in which the low level files are included, so the tags are in the import tasks of the same name. This should make it easier to understand the code and the related tags.

**Review by ja9fuchs — APPROVED** [review]
Great stuff. :+1:

**Pull request merged**